h/client_assets.py

```diff
@@ -0,0 +1,153 @@
+    return {
+      'name': '{}.html'.format(name),
+      'content': _read_resource('templates/client/{}.html'.format(name))
+    }
+
+
+def _merge(d1, d2):
+    result = d1.copy()
+    result.update(d2)
+    return result
+
+
+def websocketize(value):
+    """Convert a HTTP(S) URL into a WS(S) URL."""
+    if not (value.startswith('http://') or value.startswith('https://')):
+        raise ValueError('cannot websocketize non-HTTP URL')
+    return 'ws' + value[len('http'):]
+
+
+def asset_urls(webassets_env, name):
+    return webassets_env[name].urls()
+
+
+def _app_html_context(webassets_env,
+                     api_url,
+                     base_url,
+                     ga_tracking_id,
+                     register_url,
+                     sentry_dsn,
+                     websocket_url):
+    """
+    Returns a dict of asset URLs and contents used by the sidebar app
+    HTML tempate.
+    """
+
+    if urlparse(base_url).hostname == 'localhost':
+        ga_cookie_domain = 'none'
+    else:
+        ga_cookie_domain = 'auto'
+
+    # the serviceUrl parameter must contain a path element
+    if urlparse(base_url).path == '':
+        base_url += '/'
+
+    app_config = {
+      'apiUrl': api_url,
+      'serviceUrl': base_url,
+      'websocketUrl': websocket_url,
+    }
+
+    if sentry_dsn:
+        app_config.update({
+          'raven': {
+            'dsn': sentry_dsn,
+            'release': __version__
+          }
+        })
+
+    return {
+      'app_config': json.dumps(app_config),
+      'angular_templates': map(_angular_template_context,
+                               ANGULAR_DIRECTIVE_TEMPLATES),
+      'app_css_urls': asset_urls(webassets_env, 'app_css'),
+      'app_js_urls': asset_urls(webassets_env, 'app_js'),
+      'base_url': base_url,
+      'ga_tracking_id': ga_tracking_id,
+      'ga_cookie_domain': ga_cookie_domain,
+      'register_url': register_url,
+    }
+
+
+def render_app_html(webassets_env, base_url, api_url, websocket_url,
+                    register_url, sentry_dsn, ga_tracking_id=None, extra={}):
+    """
+    Return the HTML for the Hypothesis app page,
+    used by the sidebar, stream and single-annotation page.
+
+    :param base_url: The base URL of the Hypothesis service
+                     (eg. https://hypothes.is/)
+    :param api_url: The root URL for the Hypothesis service API
+    :param websocket_url: The WebSocket URL which the client should connect to
+    :param register_url: The URL which new users should be sent to in order to
+                         create an account
+    :param sentry_dsn: The _public_ Sentry DSN for client-side crash reporting
+    :param ga_tracking_id: The Google Analytics tracking ID
+    :param extra: A dict of optional properties specifying link tags and
+                  meta attributes to be included on the page, passed through to
+                  app.html.jinja2
+    """
+    template = Template(_read_resource('templates/app.html.jinja2'))
```

Comment:
It would probably be neater to define a module-level template environment:

```python
from jinja2 import Environment, PackageLoader
jinja_env = Environment(loader=PackageLoader(__package__, 'templates'))
```

and then load templates from it later as:

```python
jinja_env.get_template('app.html.jinja2')
```